src/casm/clexulator/Correlations.cc

```diff
@@ -613,5 +615,89 @@ std::vector<unsigned int> all_correlation_indices(Index n) {
   return corr_indices;
 }
 
+/// Gradients of Correlations
+
+Eigen::MatrixXd const Correlations::grad_correlations(DoFKey const &key) {
+  ClexParamKey paramkey;
+  clexulator::ClexParamKey corr_key(m_clexulator->param_pack().key("corr"));
+  clexulator::ClexParamKey dof_key;
+
+  if (key == "occ") {
+    paramkey =
+        m_clexulator->param_pack().key("diff/corr/" + key + "_site_func");
+    dof_key = m_clexulator->param_pack().key("occ_site_func");
+  } else {
+    paramkey = m_clexulator->param_pack().key("diff/corr/" + key + "_var");
+    dof_key = m_clexulator->param_pack().key(key + "_var");
+  }
+
+  std::string em_corr, em_dof;
+  em_corr = m_clexulator->param_pack().eval_mode(corr_key);
+  em_dof = m_clexulator->param_pack().eval_mode(dof_key);
+
+  // this const_cast is not great...
+  // but it seems like the only place passing const Clexulator is a problem and
+  // it is not actually changing clexulator before/after this function
+
+  const_cast<Clexulator &>(*m_clexulator)
+      .param_pack()
+      .set_eval_mode(corr_key, "DIFF");
+  const_cast<Clexulator &>(*m_clexulator)
+      .param_pack()
+      .set_eval_mode(dof_key, "DIFF");
```

Comment:
@bpuchala any thoughts here?

Reply:
I have no problem with it. There is at least one other place where we have to do a const_cast when evaluating the change in correlations, and as long as the const object is not changed at the end of the function it makes sense.